File: 235. Lowest Common Ancestor of a Binary Search Tree/solution2.cpp
/*
68 ms
23.1 MB

TC : O(N)
SC : O(height)
*/

/**
 * @author : SahilK-027
 * @brief : Recursive solution
*/

class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* a, TreeNode* b) {
        if(root == NULL){
            return NULL;
        }
        if(root->val < a->val && root->val < b->val){
            return lowestCommonAncestor(root->right, a,b);
        }
        if(root->val > a->val && root->val > b->val){
            return lowestCommonAncestor(root->left, a,b);
        }
        return root;
    }
};

File: 563. Binary Tree Tilt/solution2.js
/*
67 ms
46.9 MB
*/
/**
 * TC: O(n)
 * SC: O(height)
*/
/**
 * @author : SahilK-027
 * @brief : DFS
 */ 

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
function solve(root, tilt){
    if(!root){
        return 0;
    }
    let leftAns = solve(root.left, tilt);
    let rightAns = solve(root.right, tilt);

    tilt.val += Math.abs(leftAns - rightAns);

    return leftAns + rightAns + root.val;
}
var findTilt = function(root) {
    const tilt = { val: 0 };
    solve(root, tilt);
    return tilt.val;
};